File: chrome/browser/supervised_user/android/java/src/org/chromium/chrome/browser/supervised_user/SupervisedUserServicePlatformDelegate.java
// Copyright 2023 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.supervised_user;

import org.jni_zero.CalledByNative;

import org.chromium.build.annotations.NullMarked;
import org.chromium.chrome.browser.tabmodel.IncognitoTabHost;
import org.chromium.chrome.browser.tabmodel.IncognitoTabHostRegistry;

@NullMarked
class SupervisedUserServicePlatformDelegate {
    /** Closes all open incognito tabs */
    @CalledByNative
    private static void closeIncognitoTabs() {
        // TODO(https://crbug.com/429478269): Switch back to IncognitoTabHostUtils helper once
        // calling things during init is handled for us.
        for (IncognitoTabHost host : IncognitoTabHostRegistry.getInstance().getHosts()) {
            host.closeAllIncognitoTabsOnInit();
        }
    }
}

File: chrome/browser/supervised_user/supervised_user_service_platform_delegate.cc
// Copyright 2023 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/supervised_user/supervised_user_service_platform_delegate.h"
#include "chrome/browser/profiles/profile.h"
#include "chrome/browser/ui/browser_list.h"

SupervisedUserServicePlatformDelegate::SupervisedUserServicePlatformDelegate(
    Profile& profile)
    : profile_(profile) {}

void SupervisedUserServicePlatformDelegate::CloseIncognitoTabs() {
  Profile* otr_profile =
      profile_->GetPrimaryOTRProfile(/* create_if_needed =*/false);
  if (otr_profile) {
    BrowserList::CloseAllBrowsersWithIncognitoProfile(
        otr_profile, base::DoNothing(), base::DoNothing(),
        /*skip_beforeunload=*/true);
  }
}
